### ISSUE TITLE
Avoid implicitly allocating params arrays in loops

### ISSUE THREAD

**Issue by terrajobst** [dotnet/runtime#33793]
Find calls to `System.*` methods inside loops, where those methods take `params` arrays and those params arrays are being allocated, and hoist the allocation to before the loop if possible.

**Category**: Performance

**Comment by jeffhandley** [comment]
Estimates:
* Analyzer: Medium
* Fixer: Large

**Comment by carlossanlop** [comment]
@terrajobst can you provide a code example?

**Comment by carlossanlop** [comment]
- Suggested severity: Info
- Suggested category: Performance
- Suggested milestone: Future

## Examples

Let's use the `Console.WriteLine` overload that takes a `params` array:
```cs
public static class Console
{
  public static void WriteLine(string format, params object?[]? arg);
}
```

### With primitives

#### Flag

Before:
```cs
int i = 0;
while (i++ < 10)
{
    Console.WriteLine("{0}, {1}, {2}, {3}", 1, 2, 3, 4);
}
```

After:
```cs
int i = 0;
object[] arg = new object[] { 1, 2, 3, 4 };
while (i++ < 10)
{
    Console.WriteLine("{0}, {1}, {2}, {3}", arg);
}
```

#### Do not flag: An element in the passed params array has variable data
```cs 
int i = 0;
while (i++ < 10)
{
    Console.WriteLine("{0}, {1}, {2}, {3}", i, 2, 3, 4);
}
```

#### Flag

Before: Variables are used inside the loop, but they weren't modified between their creation and their use:
```cs
short a = 10, b = 20, c = 30, d = 40;
int i = 0;
while (i++ < 10)
{
    Console.WriteLine("{0}, {1}, {2}, {3}", a, b, c, d);
}
```

After:
```cs
short a = 10, b = 20, c = 30, d = 40;
int i = 0;
object[] arg = new object[] { a, b, c, d };
while (i++ < 10)
{
    Console.WriteLine("{0}, {1}, {2}, {3}", arg);
}
```


### With a class
```cs
class C
{
    private string _name;
    public C(string name) => _name = name;
    public override string ToString() => _name;
}
```

#### Flag

Before:
```cs
int i = 0;
while (i++ < 10)
{
    Console.WriteLine("Format {0}, {1}", new C("a"), new C("b"), new C("c"), new C("d"));
}
```

After: Notice the data type of the array can be preserved as `object[]`
```cs
int i = 0;
object[] arg = new object[] { new C("a"), new C("b"), new C("c"), new C("d") };
while (i++ < 10)
{
    Console.WriteLine("{0}, {1}, {2}, {3}", arg);
}
```

#### Do not flag: An element in the passed params array has variable data
```cs
int i = 0;
while (i++ < 10)
{
    Console.WriteLine("{0}, {1}, {2}, {3}", new C($"{i}"), new C("b"), new C("c"), new C("d"));
}
```
---

Things to consider:
- We need to check that the array values do not get modified between iterations.
- If the analyzer flags a case, the fixer should create the array immediately before the beginning of the loop, using the data type of the method's params array parameter.
- The new array should be named after the method parameter's name, unless a variable with that name exist, then append a number.
- CS0231: `params` is always the last parameter in a method signature, so the analyzer should only check if the last parameter is a params array.

**Comment by bartonjs** [comment]
The proposed fixes, and situations where they won't fix things, seem a bit strange.  There's value in highlighting that the array allocation is happening in a loop, but the manner in which it gets fixed (or ignored) is too variable.  So we shouldn't have a fixer at this time.  (Ideally we'll get params-span and that'll make this go away)

* Category: Performance
* Severity: Hidden

**Comment by NewellClark** [comment]
I would like to be assigned to this, please.

**Comment by Youssef1313** [comment]
I think this is ALREADY implemented, but just not yet shipped?

https://github.com/dotnet/roslyn-analyzers/blob/a46689a39bd6bbea9761961b4683bb85c84e3be8/src/PerformanceSensitiveAnalyzers/Microsoft.CodeAnalysis.PerformanceSensitiveAnalyzers.sarif#L12-L24

**Comment by huoyaoyuan** [comment]
With `params ReadOnlySpan<T>` coming, is this still valuable?
